src/librustc_lint/builtin.rs

```diff
@@ -1307,3 +1307,36 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnionsWithDropFields {
         }
     }
 }
+
+/// Lint for items marked `pub` that aren't exported to other crates
+pub struct UnexportedPub;
+
+declare_lint! {
+    UNEXPORTED_PUB,
+    Allow,
+    "`pub` items not reachable from crate root"
+}
+
+impl LintPass for UnexportedPub {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(UNEXPORTED_PUB)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnexportedPub {
+    fn check_item(&mut self, cx: &LateContext, item: &hir::Item) {
```

Comment:
`check_item` doesn't cover `pub` on impl items, foreign items and struct fields.